RFC for Session Management

rishabhpoddar opened the pull request:
It is partially complete for now. So please do not merge. I have only opened this for the purposes of easier discussion. View the [RFC here](https://github.com/blitz-js/drafts/blob/session/004_blitz_session.md). Thanks!

rishabhpoddar committed:
Initial partial draft of session management RFC

rishabhpoddar committed:
describes the more secure session flow.

rishabhpoddar committed:
removes redundant information

rishabhpoddar commented:
Hello. I have a couple of questions:
- Does the community agree with having two session flows where the user can choose between the default one and the much more secure one? According to me, this is a good idea as it also allows the framework to advertise itself as being very secure (more so than most other frameworks).
- If the answer to the above question is yes, then:
    - Please please be sure to read the third sub-point of [this](https://github.com/blitz-js/drafts/blob/session/004_blitz_session.md#session-revocation-1) section. It talks about the balance between using JWTs vs opaque tokens. Should we give the choice to a user on which one they want to use? I think we do not need to give this choice as it adds more complexity.
    - Do I need to give details on the exact algorithm being used to refresh a session for the more secure method? I ask because that get's too technical for the purposes of this RFC.
- I do not yet have a good enough mental model of how blitz manages communication between the frontend and the backend. Hence, it is hard for me to come up with how users will interact with the session module. But as a reference to how we have done it for SuperTokens, you can see our documentation for node and website SDKs, and then advice on which parts are good and not good for blitz app: [NodeJs docs](https://supertokens.io/docs/nodejs/installation), [Website SDK docs](https://supertokens.io/docs/website/installation).
- Anonymous sessions are not incorporated in these flows yet. But initial thoughts are:
  - We want to allow the developer to create a JWT based session that they can associate with anonymous users. This way, we will not use any database space (especially useful for high traffic sites with a high bounce rate).
  - Session data will be stored in the JWT. If the developer wants to associate a lot of session data with that user, such that the JWT size results in cookies being more than 4kb in size, then we want to transparently store that extra information in the database.
  - The session data in the JWT should automatically be transferred to a logged-in session. I have not thought about how this will happen exactly.
  - The signing key of this JWT is different from the one used when the user is logged in since this key cannot change in the same manner. 
- **We should have a system for cron jobs.** Most "serious" apps that I have built require at least one cron job for doing some business logic. Cron jobs will also be required for both the session flows mentioned in this RFC. Technically, we can still build it without cron jobs, but then there could be issues where a session may be in the database forever even after it has expired and that user never returns to the app.

flybayer commented:
> Does the community agree with having two session flows where the user can choose between the default one and the much more secure one? According to me, this is a good idea as it also allows the framework to advertise itself as being very secure (more so than most other frameworks).

Yes

> Please please be sure to read the third sub-point of this  section. It talks about the balance between using JWTs vs opaque tokens. Should we give the choice to a user on which one they want to use? I think we do not need to give this choice as it adds more complexity. 

For the RFC we can explain the two options, then say we choose X. And we can see if anyone complains that they want to choose between them.

In general for Blitz, if there's not a clear consensus for a certain path, we give them the option to choose between the most popular paths.

> Do I need to give details on the exact algorithm being used to refresh a session for the more secure method?

No

> I do not yet have a good enough mental model of how blitz manages communication between the frontend and the backend. Hence, it is hard for me to come up with how users will interact with the session module. But as a reference to how we have done it for SuperTokens, you can see our documentation for node and website SDKs, and then advice on which parts are good and not good for blitz app:

Yeah, we're deep in exploration mode on this part right now. Once we figure this out, I have some ideas that I'll get your input on. The Supertoken API should be fine. We'll try to make as many of those auth calls as possible inside the framework, so usually the Blitz user won't interact directly with the supertoken API.

> Anonymous sessions are not incorporated in these flows yet. But initial thoughts are:

I really like these ideas! I've never personally needed session information for anonymous users. Have you found it common to need this? Anonymous session data probably isn't essential for Blitz, but building it in would be super useful for people who do need it.

> We should have a system for cron jobs.

The default deployment path of serverless means a cron job requires a third-party service.

- Apps with a small amount of users should not require a cron job
- But for apps with a large amount of users, it's fine to need a cron job

In other words, we should minimize the need for cron jobs, especially for small/simple apps. For larger apps, it's totally fine to need extra stuff. I think this should be reasonable to implement on our side?

rishabhpoddar commented:
> I've never personally needed session information for anonymous users. Have you found it common to need this?

Yes. Quite a few use cases:
- Optimising user experience while they are logged out. For example, letting them select items to buy and then once they are on checkout, letting them log in.
- Analytics purposes
- Business logic: For example, a non logged in use is only allowed to watch 10 mins of any 2 movies after which they must log in.

We can implement sessions without cron jobs. But I don't think that adding them results in a hit to performance. Also, I think that most developers need cron jobs if they want to build a fault-tolerant app. Two examples:
- Any app that does any sort of distribution of content to other user's feed in a personalised manner. Let's say you are building a Q&A app. After a user asks a question, you need to distribute it to other's feeds. Distribution of a question is a potentially long-running algorithm. This means either we implement the whole thing in the question POST API (which will keep the user waiting for a long time on the question screen), or you use cron jobs.
- An app that has payment subscription: Here the developer will need to build a cron job that checks for payments for each user every month.

Now we could argue that these are not simple apps, but they are not rare either. For both of them, a developer using blitz will have to find and integrate their own cron job system, which may be inconvenient. 

That being said, it's not a big issue to not have cron jobs inbuilt, since that is the default for large application frameworks like Node.JS. Someone can quite easily build a library for this and others could use it. I only wanted to bring out a few more points here.

flybayer commented:
Ok that sounds good. Let's definitely have anonymous session support.

For sure any significant app will need background processing and scheduled jobs. Let's move forward with "requiring" that. Small, simple apps that don't have many sessions can just ignore this, as their DB may only need cleaned once a year or something.

rishabhpoddar commented:
This is how I am thinking of implementing it (some initial thoughts):

## Core session setup
- We will create a supertokens-blitz backend library. This will have all the core session logic and will integrate with your Blitz's DB layer (just like how other queries do). Just like how Blitz is dependant on Passport.js, it will be dependant on supertokens-blitz.
- This package will be used in a Blitz middleware to provide session verification and csrf protection.
- The middleware will also take a `boolean` for if anti-csrf check needs to be done. By default, we want this to be true for all non-GET APIs (How this is determined depends on how Blitz will convert queries to APIs internally).
-  We will need a way to specify which queries need session verification, and then those queries will automatically have this middleware associated with them.

## Session creation
- We will have to expose a function to the user to create a session. That function will set all the appropriate headers / cookies in the `resp` object and also create a `session` object in the `Context` object (so that after this is called in the API, the dev can do `ctx.session.userId` and other functions).

## Other session operations like revoking and saving data.
- All of these will be exposed via the `session` object in the `Context` object. 
- Using this, the user will be able to perform actions like `ctx.session.revoke()` or `ctx.session.getSessionData()` or `ctx.session.userId`.

## Advanced session operations like syncing session data across devices
- The user can get a unique ID for the session by using a function like `ctx.session.getSessionHandle()`.
- This unique ID can be used to revoke a session, change the session data, or other properties.
- The user can also get all `sessionHandles` given a `userId` via some API exposed to them. This way, given a userId, they can basically control any of that user's session in a query.

## Frontend token management
- On the frontend side, Blitz can use the [supertokens-website](https://github.com/supertokens/supertokens-website) package which provides interceptors for `fetch` and `axios` for managing sessions.
- This will be all internally handled by blitz.

----------------
If the above makes sense, I can write some version of this in the session RFC.

flybayer commented:
Exciting stuff! Here's my thoughts:

- We need the backend auth library to work with any db client, like if a user doesn't want to use Prisma. Perhaps the auth library can provide hooks like `createSession: sessionData => /* save session in db using any db client */` so that the user of the library has full control over the actual db access. Then we can have a `blitz-auth-prisma` plugin that abstracts those hooks for anyone using prisma.
- I'm having serious second thoughts on using Passport. It seems quite tied to Express and isn't maintained well.
- We should build the backend auth library in such a way that Blitz doesn't depend on it. I.e. someone can use a different auth library if they want. 
- I think all query and mutation RPC calls will use POST, needing CSRF. CSRF will only be needed for requests from the browser, right? I.e. using curl to execute a query will not need a CSRF token.
- I think all queries and mutations will do session verification
- For the out-of-the-box auth experience with username/password, I'm thinking to basically have a `login` mutation function that can be used from the frontend like `await login({username, password})`. Inside that mutation, we can call the auth library's "createSession" function.
- Exposing is session functions on the query/mutation context object is interesting, I hadn't thought about that. Doing this makes it a bit harder to test, I think? Will most queries/mutations need to mutate the session vs just reading static session data? I didn't think so, but I could be missing something. If it's only the login/logout mutations that need this, then I think we shouldn't add functions to context and instead manually import the auth library to make the calls that way.
- We need some way to access session information on the client, like the userId and role.

rishabhpoddar commented:
Thanks for your thoughts!

> We need the backend auth library to work with any db client, like if a user doesn't want to use Prisma. Perhaps the auth library can provide hooks like createSession: sessionData => /* save session in db using any db client */ so that the user of the library has full control over the actual db access. Then we can have a blitz-auth-prisma plugin that abstracts those hooks for anyone using prisma.

Sure. Makes sense.

> I'm having serious second thoughts on using Passport. It seems quite tied to Express and isn't maintained well.

So the main advantage of Passport is that it makes it *easier* to have logins with Google, fb etc.. Initially, we can just let the user handle email or username / password authentication (or we can build that out as well). But, we don't need Passport.js for building Blitz.

> We should build the backend auth library in such a way that Blitz doesn't depend on it. I.e. someone can use a different auth library if they want.

Makes sense. How easily it is switchable depends on where we build the default session management. If we use middlewares, then it makes it quite simple.

> I think all query and mutation RPC calls will use POST, needing CSRF. CSRF will only be needed for requests from the browser, right? I.e. using curl to execute a query will not need a CSRF token.

Yes. Only the browser needs it. However, it is also only needed when a session is required to call the API. In this case, a cURL command will need the session tokens. Furthermore, the cURL command will also need the anti-csrf token, since to the backend, it's impossible to determine if the call is from a browser or cURL command (without seriously compromising on security).

> I think all queries and mutations will do session verification

Session management is one way to do API access control. API keys are another. There is a good chance that the user may want to create queries and mutations that happen if a valid API key is present. In this case, you do not need session management for that API.

> For the out-of-the-box auth experience with username/password, I'm thinking to basically have a login mutation function that can be used from the frontend like await login({username, password}). Inside that mutation, we can call the auth library's "createSession" function.

Makes sense.

> Exposing is session functions on the query/mutation context object is interesting, I hadn't thought about that. Doing this makes it a bit harder to test, I think? Will most queries/mutations need to mutate the session vs just reading static session data? I didn't think so, but I could be missing something. If it's only the login/logout mutations that need this, then I think we shouldn't add functions to context and instead manually import the auth library to make the calls that way.

There could be APIs that need to change the data associated with the current session. We can simplify it by removing the `sessionHandle` aspects of things, and only provide `getSessionData()`, `setSessionData()`, `userId` and `revokeSession()`. If the user wants to do more advanced things with the `sessionHandle`, then they can import the library themselves.

> We need some way to access session information on the client, like the userId and role.

Oh yes! I forgot to mention this in my last comment. Thanks for reminding. The way this is done is by creating another (JWT) token that is sent in the headers and stored in the localstorage. This token will not give anyone access to the APIs. It is only meant for reading data on the frontend. We can allow the user to set what content they want to store in this token. This is also called the Open ID Connect token.
The reason we have this extra token is so that we do not need to expose the critical access token to the frontend.

flybayer commented:
> Makes sense. How easily it is switchable depends on where we build the default session management. If we use middlewares (I am aware that you had said we should not), then it makes it quite simple.

I think we should use middleware for authentication, just not authorization.

> Session management is one way to do API access control. API keys are another. There is a good chance that the user may want to create queries and mutations that happen if a valid API key is present. In this case, you do not need session management for that API.

Ah yes, API keys would be the preferred method for non-browser access.

> There could be APIs that need to change the data associated with the current session. We can simplify it by removing the `sessionHandle` aspects of things, and only provide `getSessionData()`, `setSessionData()`, `userId` and `revokeSession()`. If the user wants to do more advanced things with the `sessionHandle`, then they can import the library themselves.

- Instead of `getSessionData()`, can we automatically make the DB call to get the session data and add it to context?
- Can we also move `setSessionData()` and `revokeSession()` into the library and leave out of context? If this becomes a problem, we can add them to context later on.

rishabhpoddar commented:
> I think we should use middleware for authentication, just not authorization.

Alright.

> Instead of `getSessionData()`, can we automatically make the DB call to get the session data and add it to context?

For now, I assume the default we will be to use JWT access tokens. If so, I would recommend not doing this since this would add a db call for each API, beating the whole purpose of using a JWT. Information like userID and user role will be in the JWT itself, so those will not need to be fetched via a `getSessionData()` call. Those can be put in the context object. However, any other info that the user has set using `setSessionData()` should be fetched from the db only when the user calls `getSessionData()`.

If instead, the user decides to use opaque tokens, then we can add the `sessionData` to the context object, since a db call is required anyway.

 > Can we also move `setSessionData()` and `revokeSession()` into the library and leave out of context? If this becomes a problem, we can add them to context later on.

Yea sure.

flybayer commented:
Ok cool, that makes sense about `getSessionData()`

rishabhpoddar commented:
A small correction to the above comment has been made. I changed " If we use middlewares (I am aware that you said not to), then it makes it quite simple." to " If we use middlewares, then it makes it quite simple."

It seems that we will be using a Blitz middleware for session management.

rishabhpoddar committed:
adds init, createNewSession and verifySession for default method in RFC

rishabhpoddar commented:
Hey @flybayer . I have added some more info to the RFC.
- The `createNewSession` function needs the `res` object to set the headers and cookies. We had discussed this in the dev-auth channel on slack, and we came to the conclusion that it should be `ctx.createNewSession(..)`. A middleware will add the `createNewSession` function to all the `ctx` objects for APIs that don't require a session. 

  My thoughts on this are that I prefer if the `ctx` object has the `req` and `res` in it. I am not a fan of adding this one specific function to ctx objects since what if the user is using their own session management lib? Then this extra function will be there unnecessarily. Also, this just feels "unclean".


- The user needs to be able to express which queries need session verification. We cannot automatically apply it to all queries since then we cannot distinguish a missing, but required session from a non required session.

- We had decided not to put `sessionHandle` in the `ctx.session` object. But if we don't put those, we need some unique id in the `ctx.session` object so that, that session can be identified by the `blitz-supertokens` library for manipulation of session data. So we need to have `sessionHandle` in there.

flybayer commented:
Looking pretty good overall!

After seeing this, I think it would indeed be good for the `ctx.session` object have everything for a session. This is what you were originally proposing, right?

- `ctx.session.create()`
- `ctx.session.revoke()`
- `ctx.session.getData()`
- `ctx.session.setData()`
- `ctx.session.userId`
- `ctx.session.role`
- `ctx.session.handle`

Middleware could do something like this:
```
  return {
    session: buildSessionObject(req, res)
  }
```

**Session Verification**

It's possible my thoughts will change on this, but I think all business logic should exist inside queries/mutations, not in middleware.

- Not business logic: Session verification (extracting & verifying session from request, if any)
- Business logic: returning an error if a session is required but not found

**Note:** because we are using RPC, we'll never return specific HTTP codes for a business logic error, like a missing session. 

**So here's how I picture it fitting together:**

Session middleware:

- Extract session data from the request, if present
- Verify the session is valid, if present
- Always add `ctx.session`

Queries/Mutations:

- If a session is required, but missing, `throw AuthenticationError("You must be logged in")`
- If session exists but role is not permitted, `throw AuthorizationError()`
- The component calling the query/mutation will handle these errors appropriately.

Anonymous session:

How do these work? How I picture it working is that in middleware, if a session doesn't exist, the middleware will automatically create a new anonymous session. `ctx.session.userId === null` and `ctx.session.role === 'public'`

rishabhpoddar commented:
> After seeing this, I think it would indeed be good for the ctx.session object have everything for a session. This is what you were originally proposing, right?

Yes. This, and the session object always being there implies that certain functions would exist in certain scenarios:
- For authenticated sessions, you would not have a `create` function.
- For anonymous sessions, you would not have a `revoke` function.

> Business logic: returning an error if a session is required but not found

So there are two aspects to this (I have ignored authorisation errors for now, since those are independent to session errors):
- In case of the simple method, there are three errors:
  - General error: I think here we should throw some error to the client from the middleware itself. If we don't, we will have to communicate that with the query which over complicates things for the user. This error is thrown if the db is down (as example).
  - No session error: We can create an anonymous session here.
  - CSRF token missing / invalid error: We should throw an error to the client. 
- In case of the more secure method, there are four errors:
  - General error: Same argument as above
  - No session error: We can create an anonymous session
  - Try refresh token error: This should be handled via the middleware since it signifies that the session exists, but only the access token has expired. In this case, the frontend should call the refresh API.
  - CSRF token missing / invalid error: We should throw an error to the client. 

Does the above look good to you?

> If a session is required, but missing, throw AuthenticationError("You must be logged in")

Where in the framework will the user define if a query requires a non-anonymous session?

> If session exists but role is not permitted, throw AuthorizationError()

Where in the framework will the user define what role is required for each query?

> Anonymous session:

So I prefer having a `boolean` in the session object like `isAnonymous`. If this is `true`, then a `userId` must exist.

--------
I have made changes to the RFC code according to this discussion.

rishabhpoddar committed:
changes as per discussion https://github.com/blitz-js/drafts/pull/4\#issuecomment-612569991

flybayer commented:
> * For authenticated sessions, you would not have a `create` function.
> * For anonymous sessions, you would not have a `revoke` function.

If possible I would rather keep those functions in all the time. I.e. the session object always has the same type. This way there are less things to remember as a developer.

- I think `revoke()` on an anonymous session could simply be a noop?
- Can `create()` with an existing session revoke the current session and create a new one?
 

> * In case of the simple method, there are three errors:
>   
>   * General error: I think here we should throw some error to the client from the middleware itself. If we don't, we will have to communicate that with the query which over complicates things for the user. This error is thrown if the db is down (as example).
>   * No session error: We can create an anonymous session here.
>   * CSRF token missing / invalid error: We should throw an error to the client.
> * In case of the more secure method, there are four errors:
>   
>   * General error: Same argument as above
>   * No session error: We can create an anonymous session
>   * Try refresh token error: This should be handled via the middleware since it signifies that the session exists, but only the access token has expired. In this case, the frontend should call the refresh API.
>   * CSRF token missing / invalid error: We should throw an error to the client.
> 
> Does the above look good to you?

Yes, perfect.
 
> Where in the framework will the user define if a query requires a non-anonymous session?
> 
> Where in the framework will the user define what role is required for each query?
> 
> So I prefer having a `boolean` in the session object like `isAnonymous`. If this is `true`, then a `userId` must exist.

We can come up with a more elegant API, but essentially in your query/mutation you will check the role as shown below. We could possible use decorators on the query/mutation like `@Authorize('admin')`

```js
if (ctx.session.role !== 'admin') throw AuthorizationError()
```

From an authorization/business logic perspective, I don't care if a user is logged in or not. I only care what role they have. I don't want to have to check if a session exists and then also check if a certain role exists. If we give no-session a specific role, then we only do one check.


The actual API/library for doing authorization is still TBD, but you will define all your roles in a centralized place. Then you can use those rules anywhere in frontend or backend.

Roughly something like:

```
const rules = {
  public: ['product:read'],
  staff: ['product:manage'],
  admin: ['product:manage', 'user:manage']
}
```

We can name the anonymous role anything. I would name it `anonymous` but it's so hard to type and spell correctly! 😅 `public` conveys the same concept but is much easier to type and spell :)

rishabhpoddar commented:
> If possible I would rather keep those functions in all the time. I.e. the session object always has the same type. This way there are less things to remember as a developer.
> - I think revoke() on an anonymous session could simply be a noop?
> - Can create() with an existing session revoke the current session and create a new one?

Sure! So when we create a new session, the original one is "lost". The tokens of the original one may still be alive though.

 > We can name the anonymous role anything. I would name it anonymous but it's so hard to type and spell correctly! 😅 public conveys the same concept but is much easier to type and spell :)

Sure! We can call it `"public"` for now. And i'll update the RFC to not have the `isAnonymous` boolean. This though, will prevent a user from creating their own role called `"public"` for logged in users (which is OK in my opinion).

rishabhpoddar committed:
merges anonymous and non anonymous session types

rishabhpoddar commented:
I have also made the following changes in the RFC code snippets:
- If the session is anonymous, then the `ctx.session.userId` will be `undefined`. If the general practice in other places is to use `null` instead of `undefined`, let me know.
- Earlier, the `role` field in session would be optional (in case the user does not define a role when creating a session). However, this means that if the user had to check if the session is not anonymous, they would have to do `ctx.session.role !== undefined && ctx.session.role !== "public"`. Instead, I think it's better to have a default role (if the user does not specify the role while creating a session) for all logged in users - `"genericUser"`

rishabhpoddar committed:
adds revoking session code snippet

rishabhpoddar committed:
adds an important comment for revoking session code snippet

rishabhpoddar commented:
In certain situations, the following can happen:
- Logged in user calls a query (Q1) that requires session verification.
- The session is successfully verified, and Q1 execution starts.
- Another query (for example logout) is called while the Q1 is running
- The session for that user is revoked.
- Q1 reaches a point where it needs to call `ctx.session.getData()` and then based on this info, call `ctx.session.setData()`.
- However, since the session has been revoked, `getData` or `setData` will throw an `UNAUTHORISED` error. 

So to handle the above case, I recommend that if any query throws an `UNAUTHORISED` error, an appropriate message (session expired) be sent to the frontend automatically. This will save the user from worrying about these edge cases while writing their queries.

flybayer commented:
We're making good progress on this!

- At some point we can add ability for user to define the role name for anonymous sessions.
- Let's go with userId == null for anonymous 
- I think we should require specifying a role for `session.create()`
- Yes, if an error is thrown, it will automatically be sent to the frontend
- I think we should change the input to `session.create()` to be an object:
```ts
{
  userId: string
  role: string
  privateData?: Object
  publicData?: Object
}
```
- Along with this, I think we should change `session.setData` and `session.getData` to be `session.setPrivateData` and `session.getPrivateData` so it's very clear what data it is.

- If someone needs to change the publicData that frontend can read, how do they do that? An example would be if an admin changes the role of a person from employee to manager. How will this be role be updated in backend and frontend?

rishabhpoddar committed:
changes snippets according to https://github.com/blitz-js/drafts/pull/4\#issuecomment-613229180

rishabhpoddar commented:
I have made changes to the RFC according to your last comment.

> If someone needs to change the publicData that frontend can read, how do they do that? An example would be if an admin changes the role of a person from employee to manager. How will this be role be updated in backend and frontend?

First, `publicData` is not meant to be too dynamic. If the data you need on the frontend changes often, then you are misusing `publicData` and should make an API call to fetch that data.

`publicData` should change only when the session tokens change. This is known as session token regeneration (it's a TODO in the RFC). This implies that if `publicData` changes due to a change in the user's role, the session tokens must also change and the old ones revoked. Note that this will not destroy the session, instead, it will only force the tokens to change (along with the `publicData`). This means, the `sessionHandle` and all `privateData` will remain the same.

The above is a security best practice as well (according to OWASP session cheat sheet).

flybayer commented:
Awesome, looks great!

flybayer commented:
@rishabhpoddar what are the next steps for you on this?

rishabhpoddar commented:
@flybayer I have to complete writing the RFC first so that we are on the same page on what the user interface looks like.

Then I will start work on the `blitz-supertokens` library. I will first have to architect how the default and secure method will work together. Once that is done, I will write that in the RFC as well.

Once that is finalised, we can work together on the actual implementation of the library.

flybayer commented:
Sounds good! At some point we should post this RFC officially to the main repo. The RFC doesn't have to be fully complete to do that, but most of it should be there. 

So I guess let me know when you feel like it's good enough to post over there. Then I'll probably reorganize and reword some things and add some higher level thoughts so it's more approachable for the Blitz audience.

rishabhpoddar committed:
writes about session regenration and using session handle

rishabhpoddar commented:
I have updated the RFC. A few more points:
- When creating a session, the `publicData` should contain the `userId` and `role` of the user so that the frontend can access it. This is done by default based on the params passed to the `create` function. However, if the user sets these values (`userId` or `role`) in the `publicData` object, those values will overwrite the ones passed in the params. For example
    ```ts
    let publicData = {
        userId: "user1",
        role: "admin"
    };
    
    let session = await ctx.session.create({
        userId: "user2", role: "visitor", publicData
    });
    
    console.log(session.userId) // will be user1
    console.log(session.role) // will be admin 
    ```
- I have added `regenerate` function to the session object in the context. This is because, this function needs access to the `res` object. Please see the "Session regeneration" section in the RFC and also, please see the comments in the code snippet for that section:
    ```
    /* 
    - At this point, new tokens have been set in the response header. 
    - The old access token has not been revoked yet. This is so that in case of any failure of this API, the user doesn't get logged out. 
    - The old access tokens will be revoked when the new one is used. Ideally this requires synchronisation from the frontend when calling this API, however, that can be ignored since regeneration is a very rare situation. In the worst case (very rare), the user may get logged out if multiple, parallel calls to this API are made.
    */
    ```
- We will be using Opaque access tokens for the default method. This means that every session verify request will have to hit the database. The reason I do not want to use JWTs for the default method are:
   - People will end up using long lived JWTs - which is a bad idea
   - Since we do not have any concept of refresh tokens in the default method, we cannot go about rotating the JWT signing key thereby bottlenecking the whole system's security to safeguarding one private key.
------------
If the interface for the default implementation is OK, we can merge this into the master branch. I will continue writing the interface for the more secure method and try and make it as close to the default one as possible.

flybayer commented:
Awesome, looks great to me.

One more question on session regeneration. Does session regeneration force the end-use to log back in? Or do they stay logged in without any knowledge that their session was updated?

rishabhpoddar commented:
They stay logged in without any knowledge of their session update. 

This will happen as long as the developer does not call this query (from the frontend) multiple times, concurrently. But even if they do that, 95% of the time, the user will still stay logged in. Hence, there is an extremely low chance of the user being logged out - but in theory, it can happen.

I can get into why this logout possibility exists, but that will take a long time for me to explain here. So maybe we can get on a call for it, or else I can write about it sometime later.

--------
If it looks good, you can merge it into the master branch! Thanks

flybayer commented:
Ok perfect! I'll leave this here for now until I get some time to revise it a bit. I'm focused right now on the stuff for the alpha release :)

flybayer committed:
lots of updates

flybayer committed:
few more updates

flybayer committed:
format

flybayer committed:
tweaks

flybayer committed:
add frontend session access and security audit section

MrLeebo commented:
Is the CSRF token needed? Can't the Queries/ Mutations be protected from CSRF simply by enforcing that the content-type is application/json? Since remote `<form>` elements can't post that content type.

rishabhpoddar commented:
@MrLeebo, technically that is possible, and doing so would prevent CSRF attacks. However, this is "security by accident" - Browsers will send a preflight OPTIONS API request for `content-type: application/json`, and this will prevent CSRF if the developer has restricted the request origin domains. But if tomorrow they, by mistake, set `access-control-allow-origin: "*"` then they will open themselves to CSRF attacks without knowing so.

In my opinion, it is better to explicitly solve for CSRF either by using the sameSite flag (in cookies) or using an anti-csrf token. Given that sameSite is a relatively new feature, it's possible that end users have older browsers. Hence for now, we should stick with anti-csrf tokens.

rishabhpoddar commented:
@flybayer You mentioned the following in the RFC:
> However, **you can turn on anonymous sessions**, in which case a session will be created (and stored in the DB for the default method). They will have a full session object on which you can set `publicData` and `privateData`.

If the user sets `anonymousSessions` to `true`, and all anonymous sessions are stored in the db, it can very quickly become a problem for developers since every new website visit will add an entry to the database. Instead, what we should do is to always have `anonymousSessions: true`, but only add to the database when an API explicitly associates some `privateData` to the anonymous session.

I also think that for the initial version, we do not need to implement anonymous sessions since it's not required for many apps. This is to save on the amount of development work required. In fact, we should only build the default method initially. After our first launch, we can work on anonymous sessions and the advanced session management method.

Any thoughts?

rishabhpoddar commented:
We can allow `setPublicData` to be called for other sessions as well. This has a few implications:
- For the default method:
  - `getPublicData` (for the other sessions) from the backend will immediately start returning the changed data.
  - `getPublicData` (for the other sessions) from the frontend will start returning the changed data only after the frontend has called an API that requires session verification from the time that data changed. Most of the time this will be quick since, for most apps, the frontend calls APIs often.
- For advanced method 
  - `getPublicData` (for the other sessions) from the backend will return the changed data only after the refresh API has been called for that session.
  - `getPublicData` (for the other sessions) from the frontend will return the changed data only after the refresh API has been called for that session.
  - The advantage of calling `getPublicData` using the advanced method is that no API call is required.

------
It is important to note that the above is only true when changing the public data for device B from a request originating from device A. If changing public data for device A from a request originating from device A, then that data is reflected immediately on the frontend and backend from either of the methods.

**If this is OK, then we can add the functionality of `setPublicData` for other sessions to the RFC. Right now, it only allows to call `setPublicData` for the current session.**

rishabhpoddar commented:
@flybayer is Blitz going to allow users to have sessions across multiple sub domains? For example, Tumblr shares the same session across all tumblr pages which are their own subdomains (Is this example correct?).

I ask this cause then they will not be able to use the default session management method for doing this. The reason is cause the anti csrf token will not be available in newly visited sub-domain    resulting in the user logging out.

In the advanced method, a missing anti-csrf will lead to calling the refresh API which does not need the anti-csrf token*. The call will issue a new set of tokens, including the anti-csrf token which can be used for that sub domain only.

* If a user has made a mistake of using `access-control-allow-origin: *`, other "attacking domains" can call the refresh API for a "victim domain", but they cannot read the new set of tokens since the critical tokens are transmitted as httpOnly cookies. So this is not a security issue.

----------
The issue of not being able to use the default method for this very specific use case can be mitigated by not using anti csrf tokens and just relying on same-site cookies. Maybe eventually this could be done.

rishabhpoddar commented:
For the advanced method, most of the interface will be the same as the default method. The required changes are:
- We will need a refresh API. This can be automatically provided by Blitz on some API path, like `/auth/refresh`. It is important that this API path is unique since the refresh token will be sent only to this API. A user of Bltiz does not have to know about this.
- There will be a new type of error: `TryRefreshToken` exception. This will be thrown by any API that attempts to do session verification, but fails to do so for the following reasons:
  - Access token has expired or missing
  - Anti csrf check failed
- If a `TryRefreshToken` exception is encountered in a GET API, then the app will have to redirect the user to a page that calls the refresh API and then redirect back to the original page. Hence, we will have to ask the user to build an extra page if they want to (by default we can provide them a page that has a simple spinner). I'm not too sure how the API will look like for this in the context of Blitz, so someone else can come up with the API for this point, and write it in the "Blitz Developer Interface" section.

rishabhpoddar committed:
tiny changes to a few points

flybayer commented:
> If the user sets `anonymousSessions` to `true`, and all anonymous sessions are stored in the db, it can very quickly become a problem for developers since every new website visit will add an entry to the database. Instead, what we should do is to always have `anonymousSessions: true`, but only add to the database when an API explicitly associates some `privateData` to the anonymous session.

Ok, great idea! Then I'll remove the config option for now

> I also think that for the initial version, we do not need to implement anonymous sessions since it's not required for many apps. This is to save on the amount of development work required. In fact, we should only build the default method initially. After our first launch, we can work on anonymous sessions and the advanced session management method.

Yes, we can defer.



> We can allow `setPublicData` to be called for other sessions as well. This has a few implications:
>
> ...
>
> **If this is OK, then we can add the functionality of `setPublicData` for other sessions to the RFC. Right now, it only allows to call `setPublicData` for the current session.**

Yes, let's do this.

> @flybayer is Blitz going to allow users to have sessions across multiple sub domains? For example, Tumblr shares the same session across all tumblr pages which are their own subdomains (Is this example correct?).
> 
> ...
> 
> The issue of not being able to use the default method for this very specific use case can be mitigated by not using anti csrf tokens and just relying on same-site cookies. Maybe eventually this could be done.

For now it's fine to require advanced method for this. If we can switch to same-site cookie in the future, then that would be good.

> For the advanced method, most of the interface will be the same as the default method. The required changes are:
> 
> * We will need a refresh API. This can be automatically provided by Blitz on some API path, like `/auth/refresh`. It is important that this API path is unique since the refresh token will be sent only to this API. A user of Bltiz does not have to know about this.
> * There will be a new type of error: `TryRefreshToken` exception. This will be thrown by any API that attempts to do session verification, but fails to do so for the following reasons:
>   
>   * Access token has expired or missing
>   * Anti csrf check failed

Sounds good.

> * If a `TryRefreshToken` exception is encountered in a GET API, then the app will have to redirect the user to a page that calls the refresh API and then redirect back to the original page. Hence, we will have to ask the user to build an extra page if they want to (by default we can provide them a page that has a simple spinner). I'm not too sure how the API will look like for this in the context of Blitz, so someone else can come up with the API for this point, and write it in the "Blitz Developer Interface" section.

For pages that are only client side rendered (no SSR), can we simply make an API call to the refresh endpoint instead of a full page redirect?

SSR is off by default with the current Blitz architecture.

flybayer committed:
update anonymous session section

rishabhpoddar commented:
> For pages that are only client side rendered (no SSR), can we simply make an API call to the refresh endpoint instead of a full page redirect?

Oh yea! For client-side rendering, the API for default method and advanced method will be exactly the same. 

In that case, I would even recommend using the advanced method for everyone (who is using client-side rendering) once it has been added since it has no downsides and many advantages. Earlier I was under the impression that SSR is going to be used by default. But I'm OK either way :)

flybayer commented:
Ok, so maybe we can have advanced by default, and then people who use SSR can downgrade if they want?

In this case, what about building the advanced solution first?

rishabhpoddar commented:
> In this case, what about building the advanced solution first?

Works for me!

flybayer commented:
After considering it, I think we need to stick with the basic version as default. 

Mainly because I think Next will eventually have the new React Suspense streaming SSR renderer. This will cause server & client rendering to blend together. Every page could potentially be partially SSR. 

We could always later switch the default to advanced if it makes sense.

rishabhpoddar commented:
Alright! 

> On 6 May 2020, at 20:51, Brandon Bayer <notifications@github.com> wrote:
> 
> 
> After considering it, I think we need to stick with the basic version as default.
> 
> Mainly because I think Next will eventually have the new React Suspense streaming SSR renderer. This will cause server & client rendering to blend together. Every page could potentially be partially SSR.
> 
> We could always later switch the default to advanced if it makes sense.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/blitz-js/drafts/pull/4#issuecomment-624714032>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAWWUHZ7HTSF5NPJE2SXH6TRQF57TANCNFSM4LDY5QPA>.
>

flybayer committed:
some updates
